analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.TryCatchFinally.cs

```diff
@@ -828,9 +827,283 @@ public void TryCatch_NestedThrowWithMultipleCatchFiltersOnDifferentLevels()
+                "BeforeTry",
+                "InTry");
+        }
+
+        [TestMethod]
+        public void TryCatch_Throw_CatchAllWhen_IsUnknown()
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry"";
+    throw new FormatException();
+    tag = ""UnreachableInTry"";
+}
+catch when (arg)
+{
+    tag = ""InCatch"";
+}
+tag = ""End"";";
+            var validator = SETestContext.CreateCS(code, ", bool arg").Validator;
+            validator.ValidateTagOrder(
+                "BeforeTry",
+                "InTry",
+                "InCatch",
+                "End");
+        }
+
+        [TestMethod]
+        public void TryCatch_Throw_CatchDoesNotDowncast()
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry"";
+    throw couldBeAnything;
+    tag = ""UnreachableInTry"";
+}
+catch (FormatException)
+{
+    tag = ""UnreachableInCatch"";
+}
+tag = ""UnreachableEnd"";";
+            var validator = SETestContext.CreateCS(code, ", Exception couldBeAnything").Validator;
+            validator.ValidateTagOrder(
+                "BeforeTry",
+                "InTry");   // Signature returns Exception => we do not know that it is FormatException
+        }
+
+        [TestMethod]
+        public void TryCatch_ThrowUnexpectedException()
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry"";
+    throw new FormatException();
+    tag = ""UnreachableInTry"";
+}
+catch (NotSupportedException)
+{
+    tag = ""UnreachableInCatch"";
+}
+tag = ""Unreachable"";";
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTagOrder(
+                "BeforeTry",
+                "InTry");
+        }
+
+        [TestMethod]
+        public void TryCatch_NestedThrow_OuterCatch()
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry1"";
+    try
+    {
+        tag = ""InTry2"";
+        try
+        {
+            tag = ""InTry3"";
+            throw new FormatException();
+            tag = ""UnreachableInTry3"";
+        }
+        catch (NotSupportedException)
+        {
+            tag = ""UnreachableInCatch3"";
+        }
+        tag = ""UnreachableInTry2"";
+    }
+    catch (NotImplementedException)
+    {
+        tag = ""UnreachableInCatch2"";
+    }
+    tag = ""UnreachableInTry1"";
+}
+catch (FormatException)
+{
+    tag = ""InCatch"";
+}
+tag = ""End"";";
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTagOrder(
+                "BeforeTry",
+                "InTry1",
+                "InTry2",
+                "InTry3",
+                "InCatch",
+                "End");
+            validator.ExitStates.Should().HaveCount(1).And.ContainSingle(x => HasNoException(x));
+        }
+
+        [TestMethod]
+        public void TryCatch_NestedThrow_UnexpectedException()
```

Comment:
I think you can add one or two more tests where there are multiple catch blocks that fit the exception (e.g. DerivedException and base). In the first test the more generic bloc will be first, in the second, it should be last.